x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/xpack/security/profile/SecurityDomainIntegTests.java

```diff
@@ -40,22 +49,224 @@
+                JwtRealm.HEADER_END_USER_AUTHENTICATION,
+                JwtRealm.HEADER_END_USER_AUTHENTICATION_SCHEME + " " + HEADER_JWT_REALM_1
+            )
+        ).execute(RefreshTokenAction.INSTANCE, new CreateTokenRequest("refresh_token", null, null, null, null, refreshToken)).actionGet();
+        assertNotNull(refreshTokenResponse.getRefreshToken());
+        // "jwt" user creates token for the "file" user
+        createTokenResponse = client().filterWithHeader(
+            Map.of(
+                JwtRealm.HEADER_CLIENT_AUTHENTICATION,
+                JwtRealm.HEADER_SHARED_SECRET_AUTHENTICATION_SCHEME + " " + HEADER_SECRET_JWT_REALM_1,
+                JwtRealm.HEADER_END_USER_AUTHENTICATION,
+                JwtRealm.HEADER_END_USER_AUTHENTICATION_SCHEME + " " + HEADER_JWT_REALM_1
+            )
+        )
+            .execute(
+                CreateTokenAction.INSTANCE,
+                new CreateTokenRequest(
+                    "password",
+                    SecuritySettingsSource.TEST_USER_NAME,
+                    SecuritySettingsSourceField.TEST_PASSWORD_SECURE_STRING.clone(),
+                    null,
+                    null,
+                    null
+                )
+            )
+            .actionGet();
+        refreshToken = createTokenResponse.getRefreshToken();
+        assertNotNull(refreshToken);
+        // same-domain "index" user refreshes token
+        refreshTokenResponse = client().filterWithHeader(
+            Map.of("Authorization", basicAuthHeaderValue(RAC_USER_NAME, NATIVE_RAC_USER_PASSWORD.clone()))
+        ).execute(RefreshTokenAction.INSTANCE, new CreateTokenRequest("refresh_token", null, null, null, null, refreshToken)).actionGet();
+        assertNotNull(refreshTokenResponse.getRefreshToken());
+    }
+
+    public void testTokenRefreshFailsForUsernameOutsideDomain() {
+        // "index"-realm user creates token for the "file"-realm user
+        var createTokenResponse = client().filterWithHeader(
+            Map.of("Authorization", basicAuthHeaderValue(RAC_USER_NAME, NATIVE_RAC_USER_PASSWORD.clone()))
+        )
+            .execute(
+                CreateTokenAction.INSTANCE,
+                new CreateTokenRequest(
+                    "password",
+                    SecuritySettingsSource.TEST_USER_NAME,
+                    SecuritySettingsSourceField.TEST_PASSWORD_SECURE_STRING.clone(),
+                    null,
+                    null,
+                    null
+                )
+            )
+            .actionGet();
+        var refreshToken = createTokenResponse.getRefreshToken();
+        assertNotNull(refreshToken);
+        // fail to refresh from outside realms
+        var e = expectThrows(
+            ElasticsearchSecurityException.class,
+            () -> client().filterWithHeader(
+                Map.of(
+                    "Authorization",
+                    basicAuthHeaderValue(
+                        SecuritySettingsSource.TEST_USER_NAME,
```

Comment:
Should this be
```suggestion
                        RAC_USER_NAME,
```
to attempt refresh with the `rac_user` from the file realm that is _not_ in the same domain with native realm?

Reply:
[](https://github.com/elastic/elasticsearch/pull/85010/commits/f000848a81f2516f7f77f405eef065f58d0989ed)[f000848](https://github.com/elastic/elasticsearch/pull/85010/commits/f000848a81f2516f7f77f405eef065f58d0989ed)